File: pythonCode/faceLandmarkDetectorPolylines.py
import cv2
import dlib
import sys
import numpy as np

#draw polylines
def drawPolyline(image, landmarks, start, end, isClosed=False):
        points = []
        for i in range(start, end+1):
                point = [landmarks.part(i).x, landmarks.part(i).y]
                points.append(point)

        points = np.array(points, dtype=np.int32)
        cv2.polylines(image, [points], isClosed, (0, 255, 255), 2, 16)


def drawPolylines(image, landmarks):
        drawPolyline(image, landmarks, 0, 16)           # Jaw line
        drawPolyline(image, landmarks, 17, 21)          # Left eyebrow
        drawPolyline(image, landmarks, 22, 26)          # Right eyebrow
        drawPolyline(image, landmarks, 27, 30)          # Nose bridge
        drawPolyline(image, landmarks, 30, 35, True)    # Lower nose
        drawPolyline(image, landmarks, 36, 41, True)    # Left eye
        drawPolyline(image, landmarks, 42, 47, True)    # Right Eye
        drawPolyline(image, landmarks, 48, 59, True)    # Outer lip
        drawPolyline(image, landmarks, 60, 67, True)    # Inner lip

#Read images
imageSingle = cv2.imread("../assets/anish.jpg")
imageMultiple = cv2.imread("../assets/anish2.jpg")

#create images clone to work on
imageSingleClone = imageSingle.copy()
imageMultipleClone = imageMultiple.copy()

#convert to dlib image format
dlibImageSingle = cv2.cvtColor(imageSingleClone, cv2.COLOR_BGR2RGB)
dlibImageMultiple = cv2.cvtColor(imageMultipleClone, cv2.COLOR_BGR2RGB)

#define face detector
faceDetector = dlib.get_frontal_face_detector()

#define landmark detector and load face landmark model
landmarkDetector = dlib.shape_predictor("../dlibAndModel/shape_predictor_68_face_landmarks.dat")

#detect faces in the images
facesSingle = faceDetector(dlibImageSingle, 0)
facesMultiple = faceDetector(dlibImageMultiple, 0)

#loop over all the faces detected
for i in range(0, len(facesSingle)):
        dlibRectangle = dlib.rectangle(int(facesSingle[i].left()), int(facesSingle[i].top()), int(facesSingle[i].right()), int(facesSingle[i].bottom()))
        
        #for each face run landmark detector
        landmarks = landmarkDetector(dlibImageSingle, dlibRectangle)

        #print number of face landmarks detected
        print("number of face landmarks: ", len(landmarks.parts()))

        #draw polylines on the face
        drawPolylines(imageSingleClone, landmarks)


for i in range(0, len(facesMultiple)):
        dlibRectangle = dlib.rectangle(int(facesMultiple[i].left()), int(facesMultiple[i].top()), int(facesMultiple[i].right()), int(facesMultiple[i].bottom()))
        
        #for each face run landmark detector
        landmarks = landmarkDetector(dlibImageMultiple, dlibRectangle)

        #print number of face landmarks detected
        print("number of face landmarks: ", len(landmarks.parts()))

        #draw polylines on the face
        drawPolylines(imageMultipleClone, landmarks)

#create windows to display images
cv2.namedWindow("single person image", cv2.WINDOW_NORMAL)
cv2.namedWindow("single person landmarks", cv2.WINDOW_NORMAL)
cv2.namedWindow("multiple person image", cv2.WINDOW_NORMAL)
cv2.namedWindow("multiple person landmarks", cv2.WINDOW_NORMAL)

#display images
cv2.imshow("single person image", imageSingle)
cv2.imshow("single person landmarks", imageSingleClone)
cv2.imshow("multiple person image", imageMultiple)
cv2.imshow("multiple person landmarks", imageMultipleClone)

#press esc to exit the program
cv2.waitKey(0)

#close all the opened windows
cv2.destroyAllWindows()

File: cPlusPlusCode/faceLandmarkDetectorPolylines.cpp
#include<dlib/image_processing.h>
#include<dlib/image_processing/frontal_face_detector.h>
#include<dlib/opencv.h>
#include<opencv2/highgui.hpp>
#include<opencv2/imgproc.hpp>
#include<iostream>

using namespace dlib;
using namespace std;

//draw polyline
void drawPolyline(cv::Mat &image, full_object_detection landmarks, int start, int end, bool isClosed=false){
    std::vector<cv::Point> points;
    for(int i=start; i<=end; i++){
        points.push_back(cv::Point(landmarks.part(i).x(), landmarks.part(i).y()));
    }
    cv::polylines(image, points, isClosed, cv::Scalar(0, 255, 255), 2, 16);
}

void drawPolylines(cv::Mat &image, full_object_detection landmarks){
    drawPolyline(image, landmarks, 0, 16);              //jaw line
    drawPolyline(image, landmarks, 17, 21);             //left eyebrow
    drawPolyline(image, landmarks, 22, 26);             //right eyebrow
    drawPolyline(image, landmarks, 27, 30);             //Nose bridge
    drawPolyline(image, landmarks, 30, 35, true);       //lower nose
    drawPolyline(image, landmarks, 36, 41, true);       //left eye
    drawPolyline(image, landmarks, 42, 47, true);       //right eye
    drawPolyline(image, landmarks, 48, 59, true);       //outer lip
    drawPolyline(image, landmarks, 60, 67, true);       //inner lip
}

int main(){

    //read images
    cv::Mat imageSingle = cv::imread("../assets/anish.jpg");
    cv::Mat imageMultiple = cv::imread("../assets/anish2.jpg");

    //create images clone to work on
    cv::Mat imageSingleClone = imageSingle.clone();
    cv::Mat imageMultipleClone = imageMultiple.clone();

    //Define face detector
    frontal_face_detector faceDetector = get_frontal_face_detector();

    //define landmark detector
    shape_predictor landmarkDetector;

    //load the face landmark model
    deserialize("../dlibAndModel/shape_predictor_68_face_landmarks.dat") >> landmarkDetector;

    //convert opencv image format to dlib image format
    cv_image<bgr_pixel> dlibImageSingle(imageSingleClone);
    cv_image<bgr_pixel> dlibImageMultiple(imageMultipleClone);

    //detect faces in the image
    std::vector<rectangle> facesSingle = faceDetector(dlibImageSingle);
    std::vector<rectangle> facesMultiple = faceDetector(dlibImageMultiple);

    //loop over all detected faces
    for(int i=0; i<facesSingle.size(); i++){
        //for each face run landmark detector
        full_object_detection landmarks = landmarkDetector(dlibImageSingle, facesSingle[i]);

        //print number of landmarks detected
        cout<<"Number of face landmarks detected: "<<landmarks.num_parts()<<endl;

        //draw polyline around face landmarks
        drawPolylines(imageSingleClone, landmarks);
    }
    for(int i=0; i<facesMultiple.size(); i++){
        //for each face run landmark detector
        full_object_detection landmarks = landmarkDetector(dlibImageMultiple, facesMultiple[i]);

        //print number of landmarks detected
        cout<<"Number of face landmarks detected: "<<landmarks.num_parts()<<endl;

        //draw polyline around face landmarks
        drawPolylines(imageMultipleClone, landmarks);
    }

    //create windows to display images
    cv::namedWindow("single face", cv::WINDOW_NORMAL);
    cv::namedWindow("single face landmarks", cv::WINDOW_NORMAL);
    cv::namedWindow("multiple face", cv::WINDOW_NORMAL);
    cv::namedWindow("multiple face landmarks", cv::WINDOW_NORMAL);

    //display images
    cv::imshow("single face", imageSingle);
    cv::imshow("single face landmarks", imageSingleClone);
    cv::imshow("multiple face", imageMultiple);
    cv::imshow("multiple face landmarks", imageMultipleClone);

    //press esc to exit program
    cv::waitKey(0);

    //close all the opened windows
    cv::destroyAllWindows();

    return 0;
}